File: algo/seg_sieve.cpp
#include <iostream>
#include <vector>

using namespace std; 

void simpleSieve(vector<int> prime, int n) {
    vector<bool> res(n+1, true); 
    for (int i=2; i*i<n; i++) { 
        if (prime[i]) {
            for (int j=i; j<n; j+=i) 
                prime[i] = false; 
        }
    }

    for (int p=2; p<n; p++) {
        if (res[p])

    }
}

void segmentedSieve(int n) { 
    int limit = floor(sqrt(n)) + 1; 
    vector<int> prime;  
    prime.reserve(limit); 
    simpleSieve(limit, prime); 

    int low = limit; 
    int high = 2*limit; 

    while (low < n) { 
        if (high >= n)
            high = n; 
        
        vector<bool> res(n+1, true); 
        for (int i=0; i<n; i++) {
            int lo = floor(low/prime[i]) * prime[i]; 
            if (lo < low)
                lo += prime[i];

            for (int j=lo; j<high; j+=prime[i]) {
                res[j-lo] = false; 
            } 
        }

        for (int i=low; i<high; i++) {
            if (res[i-low] == true) {
                cout << i << " ";
            }

            low += limit; 
            high += limit;
        }
    }
}

int main() {
    int n = 100; 
    cout << "Primes smaller than " << n << endl; 
    segmentedSieve(n); 
    return 0; 
}

File: algo/sts3006/SegSieve.java
import java.util.Vector; 
import static java.lang.Math.sqrt; 
import static java.lang.Math.floor; 

class SegSieve {
    static void sieve(int limit, Vector<Integer> prime) { 
        boolean res[] = new boolean[limit+1]; 

        for (int i = 0; i < res.length; i++) { 
            res[i] = true; 
        }

        for (int p=2; p*p<limit; p++) {
            if (res[p] == true) {
                for (int i=p*p; i<=limit; i+=p) {
                    res[i] = false;
                }
            }
        }

        for (int p=2; p<limit; p++)
        {
            if (res[p] == true) {
                prime.add(p); 
                System.out.print(p + " ");
            }
        }
    }

    static void segSieve(int n) {
        int limit = (int) (floor(sqrt(n)) + 1); 
        Vector<Integer> prime = new Vector<>(); 
        sieve(limit, prime);

        int low = limit; 
        int high = 2*limit; 

        while (low < n) { 
            if (high >= n) 
                high = n; 
            
            boolean res[] = new boolean[limit+1]; 
            for (int i=0; i < res.length; i++) {
                res[i] = true; 
            }

            for (int i=0; i<prime.size(); i++) {
                int lo = (int) (floor(low/prime.get(i)) * prime.get(i));
                if (lo < low)
                    lo += prime.get(i); 

                for (int j=lo; j<high; j+= prime.get(i))
                    res[j-low] = false; 
        }

        for (int i = low; i<high; i++)
            if (res[i-low] == true)
                System.out.print(i + " ");

        low += limit; 
        high += limit; 
        }
    }   

    public static void main(String args[]) { 
        int n = 100; 
        segSieve(n); 
    }
}